70_FLIPFluids/app_resources/gridUtils.hlsl

```diff
@@ -0,0 +1,67 @@
+#ifndef _FLIP_EXAMPLE_GRID_UTILS_HLSL
+#define _FLIP_EXAMPLE_GRID_UTILS_HLSL
+
+#ifdef __HLSL_VERSION
+struct SGridData
+{
+    float gridCellSize;
+    float gridInvCellSize;
+    float pad0[2];
+
+    int4 particleInitMin;
+    int4 particleInitMax;
+    int4 particleInitSize;
+
+    float4 worldMin;
+    float4 worldMax;
+    int4 gridSize;
+};
+
+static const float POSITION_EPSILON = 1e-4;
+
+float4 clampPosition(float4 position, float4 gridMin, float4 gridMax)
+{
+    return float4(clamp(position.xyz, gridMin.xyz + POSITION_EPSILON, gridMax.xyz - POSITION_EPSILON), 1);
+}
+
+inline uint cellIdxToFlatIdx(int3 index, int4 gridSize)
+{
+    uint3 idxClamp = clamp(index, (int3)0, gridSize.xyz - 1);
+    return idxClamp.x + idxClamp.y * gridSize.x + idxClamp.z * gridSize.x * gridSize.y;
+}
+
+inline int3 flatIdxToCellIdx(uint id, int4 gridSize)
+{
+    int x = id % gridSize.x;
+    int y = id / gridSize.x % gridSize.y;
+    int z = id / (gridSize.x * gridSize.y);
+    return int3(x, y, z);
+}
```

Comment:
Here's the thing about using snake curve to 3D space fill:
1. bad spatial locality
2. integer divisions are super slow
3. the compiler doesn't see `gridSize` as a constant so it can't optimize/precomp your integer divs

If you ever need to space-fill and putting a constraint that the grid size must be a power-of-two is an option, then use a Morton Code (Z-order).


Either way I think you should just use 3D images for everything and never need to go 1D -> 3D ever.


P.S. going from MAJOR to minor, skips all `%` in a snake-curve addressing func
```cpp
z = id/gridSliceStride; // precomputed `gridSize.x*gridSize.y
tmp = id-z*gridSliceStride;
y = tmp/gridSize.x;
x = tmp-y*gridSize.x;
```

Reply:
no need to make it more efficient... as I kinda expect you to not use these mappings